File: 01-week-1--starter-algorithms/03-day-4--selection-sort/ruby/selection_sort.rb
def selection_sort(arr)
  sorted_array = []
  until arr.length == 0
    num = arr.min
    sorted_array.push(num)
    indexedNum = arr.index(num)
    arr.delete_at(indexedNum)
  end
  return sorted_array
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort([3, -1, 5, 2])

  puts

  # Don't forget to add your own!

  # BENCHMARK HERE, you can print the average runtime
  long_input = []

  100.times { long_input << rand }
end

# Please add your pseudocode to this file
# And a written explanation of your solution

# sort an input array from min to max
# create an empty variable to store new array
# iterate through array as long as its not empty
# destructively remove the smallest element that we iterated on
# return the new sorted array

#what did WE DO?????
#we needed an empty variable to store our new sorted array
#we had to iterate through the array until length hit 0- bc we progressively removed things from array
# we removed things by finding the index the min number was at.
#if we didnt remove things from the array, we would never break out the loop
#everytime we found the min, we pushed it into the new array, and removed it from input array

File: 01-week-1--starter-algorithms/03-day-4--selection-sort/javascript/selection_sort.js
function selectionSort(arr) {
  let sortedAry= []
    while (arr.length > 0){
    let num = Math.min(...arr)
    //WHY DOES IT NEED A SPREAD OPERATOR??????
    // ...arr makes a new copy you can work with 
    // initial array doesnt exist anymore. but you can create a new copy that can exist now
    let newSplice= arr.indexOf(num)
    sortedAry.push(num)
    arr.splice(newSplice,1)
    }
    return sortedAry

  // type your code here
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSort([3, -1, 5, 2]));

  console.log("");

  // BENCHMARK HERE, and print the average runtime
  const longInput = [];

  for (let i = 0; i < 100; ++i) {
    longInput.push(Math.random());
  }
}

module.exports = selectionSort;

// Please add your pseudocode to this file
//sorting an array with selection sort
//create a new variable that's the empty array/ output
// find out the minimum with Math.min()- this gives the input of the array at the smollest val
// take out the value that was added 



// And a written explanation of your solution
//we needed an empty variable to store all the min values we took from the input array
// we had to iterate through the input array, and grab the min value from it each time.
//we had to work with a copy of the input array instead of the original array, in our math min function bc we were
//destructively changing the array when removing values from the input array
//we removed vals from the input array so that we dont get dupes